### PR TITLE
Fixes for PQC test cases and Fix Serialization issues with keys.

### PR DESCRIPTION
**Pull request by johnpeck-us-ibm**
Needed to remove the dependency of JDK17 in the testcases.
Also, there is an issue with serialization. When KeyRep is used to serialize the key the provider order is used to de serialize the key which can cause the wrong key to be created.  So, added a new KeyRep class that contains the provider which will allow the correct key to be created.

### CODE REVIEW

File: src/main/java/com/ibm/crypto/plus/provider/JCEPlusKeyRep.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import java.io.NotSerializableException;
+import java.io.ObjectStreamException;
+import java.io.Serializable;
+import java.security.KeyFactory;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Locale;
+import javax.crypto.spec.SecretKeySpec;
+
+class JCEPlusKeyRep implements Serializable {
```
Comment (jasonkatonica) [inline src/main/java/com/ibm/crypto/plus/provider/JCEPlusKeyRep.java:20]: This class has mostly the same code as `java.security.KeyRep` could we instead extended that class to add the provider which seems like the main change here.

File: src/main/java/com/ibm/crypto/plus/provider/DHPublicKey.java [hunk 2]
```diff
@@ -394,7 +393,7 @@ private void checkDestroyed() {
      */
     private Object writeReplace() throws java.io.ObjectStreamException {
         checkDestroyed();
-        return new KeyRep(KeyRep.Type.PUBLIC, getAlgorithm(), getFormat(), getEncoded());
+        return new JCEPlusKeyRep(JCEPlusKeyRep.Type.PUBLIC, getAlgorithm(), getFormat(), getEncoded(), provider.getName());
```
Comment (jasonkatonica) [inline src/main/java/com/ibm/crypto/plus/provider/DHPublicKey.java:396]: Will this cause compatibility problems with previous keys that have been serialized directory to disk and or stored in a keystore file? Id assume since we are now changing what object is written to disk ( JCEPlusKeyRep instead of KeyRep ) then serialized objects from the past would not be compatible. Similar comment for other key classes updated.

I can see how tracking the provider in the serialized object will allow us to ignore the provider order when we deserialize from disk.

**Review by jasonkatonica — COMMENTED** [review]
There are some conflicts in some files in this PR in addition to a few comments here. We should rebase this on latest code to resolve the conflicts. 

On other general comment is coverage for testing here. I do see a serilaization test for PQC keys here however not too many other tests. I am OK either way since its the same exact pattern of using the same class to serialze but ill mention it in case there are any other corner cases here that we want to test with other keys that dont follow the same pattern.....

File: src/main/java/com/ibm/crypto/plus/provider/JCEPlusKeyRep.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                    keyType + " " + encodingFormat);
+            }
+        } catch (NotSerializableException nse) {
+            throw nse;
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+            NotSerializableException nse = new NotSerializableException("java.security.Key: " +
+                "[" + keyType + "] " +
+                "[" + keyAlg + "] " +
+                "[" + encodingFormat + "]");
```
[the remark this answers is not in the document]
Comment (johnpeck-us-ibm) [reply src/main/java/com/ibm/crypto/plus/provider/JCEPlusKeyRep.java]: Added Provider to this message.

**Comment by jasonkatonica** [comment]
The code coverage indicated that the serilization methods are not being executed. Perhaps this is due to the new tests not added to `TestAll.java` here is a view of PQCPublicKey which does not seem to show any coverage there:

image

**Comment by johnpeck-us-ibm** [comment]
> The code coverage indicated that the serilization methods are not being executed. Perhaps this is due to the new tests not added to `TestAll.java` here is a view of PQCPublicKey which does not seem to show any coverage there:
> image

Yes, this test was missing from TestAll. I have now added this. I am not sure if we really need to add tests for other key types or not. We have not had these in the past. I will check with Karthik to see if they are testing the othe key types.

File: src/test/java/ibm/jceplus/junit/base/BaseTestKeySerialization.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            throw e;
+        }
+
+
+        // serialize and de-serialize the public key while storing in file
+        File keyFile = createTempFile();
+        serializeKey(key, keyFile);
+        SecretKey deserializedKey = (SecretKey) deserializeKey(keyFile);
+
+        /*assertEquals(key.getClass().getName(),
```
[the remark this answers is not in the document]
Comment (johnpeck-us-ibm) [reply src/test/java/ibm/jceplus/junit/base/BaseTestKeySerialization.java:113]: Removed.

File: src/test/java/ibm/jceplus/junit/base/BaseTestKeySerialization.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        // private key serialize and de-serialize
+        File privateKeyFile = createTempFile();
+        serializeKey(keyPair.getPrivate(), privateKeyFile);
+        PrivateKey deserializedPrivateKey = (PrivateKey) deserializeKey(privateKeyFile);
+
+        assertEquals(keyPair.getPublic().getClass().getName(), 
+            deserializedPublicKey.getClass().getName(),
+            "Public deserialized class does not match original");
+
+        assertEquals(keyPair.getPublic().getClass().getName(),
```
[the remark this answers is not in the document]
Comment (johnpeck-us-ibm) [reply src/test/java/ibm/jceplus/junit/base/BaseTestKeySerialization.java:70]: Fixed

File: src/test/java/ibm/jceplus/junit/base/BaseTestPQCKeySerialization.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        // private key serialize and de-serialize
+        File privateKeyFile = createTempFile();
+        serializeKey(keyPair.getPrivate(), privateKeyFile);
+        PrivateKey deserializedPrivateKey = (PrivateKey) deserializeKey(privateKeyFile);
+
+        assertEquals(keyPair.getPublic().getClass().getName(), 
+            deserializedPublicKey.getClass().getName(),
+            "Public deserialized class does not match original");
+
+        assertEquals(keyPair.getPublic().getClass().getName(),
```
[the remark this answers is not in the document]
Comment (KostasTsiounis) [reply src/test/java/ibm/jceplus/junit/base/BaseTestPQCKeySerialization.java:66]: This doesn't seem to have been fixed with the last push.

File: src/test/java/ibm/jceplus/junit/openjceplus/TestAll.java [hunk 7]
```diff
@@ -104,6 +104,7 @@
     TestPQCKeyInteropBC.class,
     TestPQCKeyInteropOracle.class,
     TestPQCKeys.class,
+    TestPQCKeySerialization.class,
```
[the remark this answers is not in the document]
Comment (jasonkatonica) [reply src/test/java/ibm/jceplus/junit/openjceplus/TestAll.java:108]: OK makes sense.

File: src/test/java/ibm/jceplus/junit/base/BaseTestPQCKeySerialization.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (johnpeck-us-ibm) [reply src/test/java/ibm/jceplus/junit/base/BaseTestPQCKeySerialization.java:66]: Your right. I got it now.

File: src/test/java/ibm/jceplus/junit/base/BaseTestPQCKeySerialization.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import javax.crypto.KEM;
+import javax.crypto.SecretKey;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class BaseTestPQCKeySerialization extends BaseTestJunit5 {
```
[the remark this answers is not in the document]
Comment (jasonkatonica) [reply src/test/java/ibm/jceplus/junit/base/BaseTestPQCKeySerialization.java]: LGTM

**Review by jasonkatonica — APPROVED** [review]
LGTM
